### PR TITLE
Added "error" event listener

### LINKED ISSUE #2561: Cannot catch EADDRINUSE from server.start()
When I try to call `server.start(callback)`, and the port is in use, the script crashes with a message:

```
events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: listen EADDRINUSE
    at errnoException (net.js:904:11)
    at Server._listen2 (net.js:1042:14)
    at listen (net.js:1064:10)
    at net.js:1146:9
    at dns.js:72:18
    at process._tickCallback (node.js:419:13)
```

I would expect the `callback` to be called instead (with the error, of course).

### PR DESCRIPTION
**Pull request by arb**
Added error listener to listener object. Closes #2561.

### CODE REVIEW

**Comment by kanongil** [comment]
This approach can cause the callback to be triggered multiple times if an error is triggered on the listener for some other reason at a date after the `listening` event has been triggered.

You should probably also change `Connection#info.started` to `0`, to indicate that it stopped running.

**Comment by arb** [comment]
@kanongil Yeah I knew :-( But I think I have a solution that isn't vulnerable to that. I also though about trying to call `_stop` on error but I'm not sure if we really need to do that.

**Comment by Marsup** [comment]
Why don't you unregister that listener once the server is started ?

File: lib/connection.js [hunk 1]
```diff
@@ -149,12 +149,26 @@ internals.Connection.prototype._start = function (callback) {
         return process.nextTick(callback);
     }
 
+    var self = this;
+    var startupErr = function (err) {
```
Comment (hueniverse) [inline lib/connection.js]: Event handler should be called `onError`

File: lib/connection.js [hunk 2]
```diff
@@ -149,12 +149,26 @@ internals.Connection.prototype._start = function (callback) {
...
+
+        self._started = false;
+        return callback(err);
+    };
+    this.listener.once('error', startupErr);
+
+    var afterListen = function () {
```
Comment (hueniverse) [inline lib/connection.js]: This is usually called `finalize`

File: test/connection.js [hunk 3]
```diff
@@ -399,6 +399,20 @@ describe('Connection', function () {
                 });
             });
         });
+
+        it('will return an error if the port is aleady in use', function (done) {
+
+            var server = new Hapi.Server();
+            server.connection({ port: 9000 });
```
Comment (hueniverse) [inline test/connection.js]: test should first pass `0` (or nothing) and then reuse the port assigned.

File: lib/connection.js [hunk 4]
```diff
@@ -149,12 +149,26 @@ internals.Connection.prototype._start = function (callback) {
         return process.nextTick(callback);
     }
 
+    var self = this;
```
Comment (hueniverse) [inline lib/connection.js]: `self` always at top of function with empty line.

File: lib/connection.js [hunk 5]
```diff
@@ -149,12 +149,26 @@ internals.Connection.prototype._start = function (callback) {
...
+    var onError = function (err) {
+
+        self._started = false;
+        return callback(err);
+    };
+    this.listener.once('error', onError);
```
Comment (hueniverse) [inline lib/connection.js]: While at it, add another empty line above this :-)

**Pull request merged**